tools/ci.sh

```diff
@@ -93,33 +93,38 @@ function ci_code_size_build {
     echo "Comparing sizes of reference ${REFERENCE} to ${COMPARISON}..."
     git log --oneline $REFERENCE..$COMPARISON
 
-    function code_size_build_step {
-        COMMIT=$1
-        OUTFILE=$2
+    OLD_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
 
-        echo "Building ${COMMIT}..."
-        git checkout --detach $COMMIT
-        git submodule update --init $SUBMODULES
-        git show -s
-        tools/metrics.py clean $PORTS_TO_CHECK
-        tools/metrics.py build $PORTS_TO_CHECK | tee $OUTFILE
-        return $?
-    }
+    ( # Execute in a subshell so the trap & code_size_build_step doesn't leak
+        function code_size_build_step {
+            if [ ! -z "$OLD_BRANCH" ]; then
+                trap 'git checkout "$OLD_BRANCH"' RETURN EXIT ERR
+            fi
 
-    # Allow errors from tools/metrics.py to propagate out of the pipe above.
-    set -o pipefail
+            COMMIT=$1
+            OUTFILE=$2
 
-    # build reference, save to size0
-    # ignore any errors with this build, in case master is failing
-    code_size_build_step $REFERENCE ~/size0
-    # build PR/branch, save to size1
-    code_size_build_step $COMPARISON ~/size1
-    STATUS=$?
+            echo "Building ${COMMIT}..."
+            git checkout --detach $COMMIT
+            git submodule update --init $SUBMODULES
+            git show -s
+            tools/metrics.py clean "$PORTS_TO_CHECK"
+            # Allow errors from tools/metrics.py to propagate out of the pipe below.
+            set -o pipefail
+            tools/metrics.py build "$PORTS_TO_CHECK" | tee $OUTFILE
+        }
 
-    set +o pipefail
-    unset -f code_size_build_step
+        # build reference, save to size0
+        # ignore any errors with this build, in case master is failing
+        code_size_build_step $REFERENCE ~/size0
+        # build PR/branch, save to size1
+        code_size_build_step $COMPARISON ~/size1
```

Comment:
Is the return value of this last step propagated out as the return value / result of the entire `ci_code_size_build` function?

Reply:
We're running in `set -e` so any individual nonzero exit status not otherwise handled makes the whole thing exit. So if build_step fails (whether for REFERENCE or COMPARISON) the whole job fails.

In my local WIP I've restored $IGNORE_ERRORS, which means that a failure at REFERENCE results in success (zero exit status), while a failure at COMPARISON results in success (nonzero exit status)

```
(micropython) jepler@bert:~/src/micropython$ PORTS_TO_CHECK=bm REFERENCE=HEAD COMPARISON=broken-minimal bash -x ./tools/ci.sh code_size_build > /dev/null 2>&1 ; echo $?
1
(micropython) jepler@bert:~/src/micropython$ PORTS_TO_CHECK=bm REFERENCE=broken-minimal COMPARISON=HEAD bash -x ./tools/ci.sh code_size_build > /dev/null 2>&1 ; echo $?
0
```